internal/database/types_hcpcluster.go

```diff
@@ -0,0 +1,139 @@
+package database
+
+import (
+	"strings"
+	"encoding/json"
+
+	"github.com/google/uuid"
+	"github.com/Azure/ARO-HCP/internal/api"
+	"github.com/Azure/ARO-HCP/internal/api/arm"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	azcorearm "github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
+)
+
+type BaseDocument struct {
+	ID string `json:"id,omitempty"`
+
+	// https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/time-to-live
+	TimeToLive int `json:"ttl,omitempty"`
+
+	// System-defined properties generated by Cosmos DB
+	CosmosResourceID  string      `json:"_rid,omitempty"`
+	CosmosSelf        string      `json:"_self,omitempty"`
+	CosmosETag        azcore.ETag `json:"_etag,omitempty"`
+	CosmosAttachments string      `json:"_attachments,omitempty"`
+	CosmosTimestamp   int         `json:"_ts,omitempty"`
+}
+
+// newBaseDocument returns a baseDocument with a unique ID.
+func NewBaseDocument() BaseDocument {
+	return BaseDocument{ID: uuid.New().String()}
+}
+
+type TypedDocument struct {
+	BaseDocument
+	PartitionKey string          `json:"partitionKey"`
+	ResourceType string          `json:"resourceType"`
+	Properties   json.RawMessage `json:"properties"`
+}
+
+// newTypedDocument returns a TypedDocument from a ResourceType.
+func NewTypedDocument(partitionKey string, resourceType azcorearm.ResourceType) *TypedDocument {
+	return &TypedDocument{
+		BaseDocument: NewBaseDocument(),
+		PartitionKey: strings.ToLower(partitionKey),
+		ResourceType: strings.ToLower(resourceType.String()),
+	}
+}
+
+// HCPClusterDocument represents a customer desired HCP Cluster.
+// To transition from our current state using cluster-service as half the source of truth to a state where
+// cosmos contains all the desired state and all the observed state, we are basing the schema on ResourceDocument.
+type HCPClusterDocument struct {
+	// do NOT marshal this field.  These are system properties that we will set on read so we can look at them, but the values
+	// in this type are never serialized because the typed document is the wrapper, not the content.
+	TypedDocument    `json:"-"`
```

Comment:
This is present so we have etag information. Hiding it makes integration with the wider cosmos ecosystem harder and doesn't benefit us since we can remove it from serialization here.